cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+a list of rules into a scheme to attempt to match rules in some order,
+possibly sharing match operations between similar rules to reduce
+work.
+
+The core data structure we produce is a "decision trie" per internal
+constructor body. This is an intermediate representation of sorts that
+is built from individual-rule IR (LHS + RHS) sequences, and is then
+used to generate Rust source.
+
+The decision trie is, as the name implies, a kind of decision tree, in
+the sense that we start at the root and move down the tree based on
+the result of match operations (each feeding one "decision").
+
+It is a "trie" (which is a kind of tree) because at each level, its
+edges are labeled with match operations; a trie is a tree where one
+input character from an alphabet is used to index children at each
+level.
+
+Each node in the tree is either an internal decision node, or a leaf
+"expression" node (which we reach once we have a successful rule
+match). The "execution semantics" of the trie are
+backtracking-based. We attempt to find some path down the tree through
+edges whose match ops run successfully; when we do this to reach a
+leaf, we have the values generated by all of the match ops, and we can
+execute the sequence of "expression instructions" in the leaf. Each
+rule's left-hand side becomes a series of edges (merged into the
+existing tree as we process rules) and each rule's right-hand side
+becomes one leaf node with expression instructions.
+
+At any point, if a match op does not succeed, we try the next out-edge
+in sequence. If we have tried all out-edges from a decision node and
+none were successful, then we backtrack one level further. Thus, we
+simply perform an in-order tree traversal and find the first
+successful match.
+
+Though this sounds possibly very inefficient if some decision node has
+a high fan-out, in practice it is not because the edges are often
+known to be *mutually exclusive*. The canonical example of this is
+when an enum-typed value is destructured into different variants by
+various edges; we can use a Rust `match` statement in the generated
+source and have `O(1)` (or close to it) cost for the dispatch at this
+level.
+
+Building the trie is a somewhat subtle procedure because we must be
+mindful of *priorities*. In order to ensure that rule priorities are
+obeyed, we actually label each edge with a *priority range* and a
+match op. Edges must then be considered in priority order. When
+building the trie, we uphold an important *non-overlapping-priority*
+invariant: there must not be two edges with overlapping priority
+ranges and different match ops, or else it would be unclear which to
+run next. Priority ranges are inclusive in a discrete (integer) space,
+so "overlap" is also subtle: we can allow edges with "unit" priority
+ranges (just one priority), and these do not overlap. So, e.g., we can
+allow an edge with priority `[0, 10]` (higher is more prioritized) and
```

Comment:
Would this be disallowed if the middle range were `[0, 1]`? Where in the system would a failure happen if this were attempted?

Reply:
Yes, we would avoid building a trie like that, but it wouldn't cause a failure, but rather a tree-split. The block comment in `trie.rs` covers this a bit better than this summary here (which I removed in favor of just pointing to the block comment).